### PR TITLE
1.48x Faster Partial Caching - Alternative to #36310

### PR DESCRIPTION
**Pull request by schneems**
This is a "safer" but slower alternative to #36310.

```
Then: 5.97/3.29 => 1.81x faster
Now: 5.97/4.02 => 1.48x faster
```

So the change in performance from #36310 to this PR is  about

```
1 - 3.29/4.02 => 18% # slower than #36310 but still 1.48x faster than master.
```

In #36310 performance was the main goal. In this PR the primary goal was to change as little as possible, performance was the secondary goal. The main difference between the two PRs comes from how `cache_version` and `cache_key` are computed. In #36310 that logic was ported over to the `Cache::Key` class where it could be optimized, in this PR the logic was not moved and instead the `Cache::Key` class leverages the `expanded_version` method in their original location.

How does this PR improve performance for rendering a partial and passing in a collection? 

- On master when a key is used in the cache it has to be generated 3 different times. By using the `Cache::Key` class, it is only generated once and then re-used
- On master, the key that represents the view partial that is being rendered is calculated for every object being pulled from cache inside of the loop, in this PR that view key is generated once outside the loop and re-used.

For safety: In addition to not changing the location of the `expanded_version` and `expanded_key` logic, this class will respect any modified logic of the `Rails.cache` store. So if a cache store customizes their version of `expanded_version` then that version will be used when generating the cache.

Here's some raw numbers from the benchmark provided in #36310 when applied to this PR.

```
$ bundle exec ruby perf_scripts/partial_cache.rb
-- create_table(:customers, {:force=>true})
   -> 0.0115s
/Users/rschneeman/Documents/projects/codetriage/perf_scripts/partial_cache
<#ActiveSupport::Cache::MemoryStore entries=1000, size=352893, options={}>
Warming up --------------------------------------
collection render: no cache
                        20.000  i/100ms
collection render: with cache
                         4.000  i/100ms
Calculating -------------------------------------
collection render: no cache
                        199.729  (± 9.0%) i/s -      1.000k in   5.051812s
collection render: with cache
                         49.642  (± 4.0%) i/s -    248.000  in   5.008595s

Comparison:
collection render: no cache:      199.7 i/s
collection render: with cache:       49.6 i/s - 4.02x  slower
```

> Note that while it's still slower, the prior version was 5.97x slower.

### CODE REVIEW

File: activesupport/lib/active_support/cache/key.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+        if @cache_version == "/"
+          nil
+        else
+          @cache_version
+        end
+      end
+
+      def cache_key_with_version
+        @cache_method_obj.send(:retrieve_cache_key, @key_parts)
```
Comment (kaspth) [inline activesupport/lib/active_support/cache/key.rb]: Hm, the `send`s in here bother me, but I think we can live with them if the future design idea is to flip the relationship such that a cache store really consults the key for `retrieve_cache_key` etc.

**Review by kaspth — COMMENTED** [review]
Hey Richard, really cool that you went the extra length of restructuring to slim your PR, especially when we we're negligent in reviewing. I'm sorry it took so long and thank you for your patience ❤️🙏

Though now I'm throwing a lot at you, so if you'd rather that I address my concerns in commits after this PR, then I'll happily merge now and look into them. Or we can do a bit of both, or just the usual, your pick 🤠

File: activesupport/lib/active_support/cache/key.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+      def update(key)
+        @key_parts << key
+        @cache_key = nil
+        @cache_version = nil
+        self
+      end
+      alias :<< :update
+
+      def length
```
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/lib/active_support/cache/key.rb]: I needed this in #36310 due to some tests that were asserting based on key objects since in that PR the Cache::Key objects ended up being passed around instead of only in this one collection cache context. 

I think we don't need it in this PR, I'll remove it and watch CI.

File: activesupport/lib/active_support/cache/key.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+      def cache_key
+        @cache_key ||= @cache_method_obj.send(:expanded_key, @key_parts)
+      end
+
+      def cache_version
+        @cache_version ||= @cache_method_obj.send(:expanded_version, @key_parts)
+
+        if @cache_version == "/"
+          nil
```
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/lib/active_support/cache/key.rb]: Without this code we get three failures in the tests:

```
CachedCollectionViewRenderTest#test_collection_caching_with_partial_that_doesn't_use_fragment_caching [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:797]:
Expected: "Cached"
  Actual: "Hello: david"


rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:791

.................................F

Failure:
CachedCollectionViewRenderTest#test_collection_caching_with_cached_true [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:807]:
--- expected
+++ actual
@@ -1 +1,2 @@
-"Cached"
+"  Hello: david
+"



rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:801

..F

Failure:
CachedCollectionViewRenderTest#test_template_body_written_to_cache [/Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:778]:
Expected: "Hello: david"
  Actual: nil


rails test Users/rschneeman/Documents/projects/rails/actionview/test/template/render_test.rb:773
```

Here's an example of one of the tests:

```ruby
  test "collection caching with cached true" do
    customer = CachedCustomer.new("david", 1)
    key = cache_key(customer, "test/_cached_customer")

    ActionView::PartialRenderer.collection_cache.write(key, "Cached")

    assert_equal "Cached",
      @view.render(partial: "test/cached_customer", collection: [customer], cached: true)
  end
```

The issue at heart is that `nil` and empty values weren't really taken into account when the current `noramlize_version` was written so the behavior is somewhat undefined (or rather it's unclear what behavior we _want_).

For example if you try to generate a cache version based off of a symbol you'll get a nil, but if you try to generate it off of a symbol in an array you'll get an empty string:

```ruby
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, :views)
=> nil
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [:views])
=> ""
```

These tests are failing because of the base assumption that normalize version will treat `[key]` the same as `key`. This is mostly true, but not in the case of empty values. In these failures we're generating a cache_version when we wouldn't have originally

Basically the difference between:

```
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [[:views], "value", []])
# => "/"
```

and 

```
ActiveSupport::Cache::MemoryStore.new.send(:normalize_version, [:views, "value", []])
# => ""
```

My initial solution was to consider "/" the same as an empty string. I'm looking into seeing if there's another way to generate or insert keys that doesn't require this explicit knowledge.

File: activesupport/lib/active_support/cache/key.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/lib/active_support/cache/key.rb]: CI works with this code removed, going to resolve this comment.

File: activesupport/lib/active_support/cache/key.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/lib/active_support/cache/key.rb]: I've got a solution here https://github.com/rails/rails/pull/37301/commits/a01a94568cd38ceb40bd27d5ef8771565607b193

Essentially we can avoid this issue by making sure that items are directly added to the `Cache::Key`, so rather than adding a whole array to the key, we can add it's constituent parts. This method has the benefit of removing an array allocation from the internal loop.

File: activesupport/lib/active_support/cache/key.rb (hunk 1 above)
Comment (schneems) [reply activesupport/lib/active_support/cache/key.rb]: > if the future design idea is to flip the relationship such that a cache store really consults the key for retrieve_cache_key etc.

Inverting that relationship is the core idea with #36310 where the key logic would live in the object rather than in external recursive methods (i.e. `normalize_key` and `normalize_version`). It might be worth re-visiting it now if the eventual goal would be to get there. I could make it less invasive and restrict the key generation to only view collection caching. 

> Hm, the sends in here bother me

I agree that the sends here are unnerving. The interface provided by Cache::Store is a bit odd since most people who implement a cache store inherit from this store meaning they can freely use these methods even though they're written as being private. I could make them public (because in a very real way they're already public today) however that wouldn't fix any issues with other cache stores that have implemented their own methods and decided to keep them private.

This also does highlight probably the most "unsafe" thing about this PR versus #36310, it assumes any cache store provided has these 3 methods on them. To my knowledge, the vast majority of people use the Rails provided cache stores and some use `:dalli`. I'm not sure if there are any other cache stores to consider.

File: activesupport/lib/active_support/cache.rb [hunk 4]
```diff
@@ -851,3 +851,4 @@ def uncompress(value)
     end
   end
 end
+require "active_support/cache/key"
```
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/lib/active_support/cache.rb]: Good point. I fixed this so the key is lazily generated and memoized to the class in a way that doesn't require this circular dependency.

File: activesupport/lib/active_support/cache/key.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
```
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/lib/active_support/cache/key.rb]: I reduced the docs to one example. People can view the tests for more if the want/need. These docs are less needed as this class is much smaller than #36310, it's easier to grok the purpose of this class at a glance versus that PR.

File: activesupport/lib/active_support/cache/key.rb [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+      @@has_rails_cache = :unknown
+
+      def initialize(key = DEFAULT_KEY)
+        @cache_key = nil
+        @cache_version = nil
+        @key_parts = []
+        @cache_method_obj = cache_method_obj
+
+        self << key unless key == DEFAULT_KEY
+      end
```
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/lib/active_support/cache/key.rb]: I realized that we're always initializing with a "base" key so I changed it to a required argument and avoided calling the `<<` method altogether. As the class is internal only for now (:nodoc:) I think it's okay for the object to be more opinionated about how it's used. for now. Let me know if you disagree or you're worried. I could add a builder method. 

I do like the way that `Key.chain(thing)` reads in code.

File: activesupport/lib/active_support/cache/key.rb [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+          if @@has_rails_cache == :unknown
+            @@has_rails_cache = defined?(Rails) && Rails.respond_to?(:cache)
+          end
+
+          if @@has_rails_cache
+            Rails.cache
+          else
+            CACHE_METHOD_OBJ
+          end
+        end
```
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/lib/active_support/cache/key.rb]: I like using dependency injection, switched to an optional second argument that defaults to grabbing the information from the class.

I tried to inject it in `collection_caching.rb` directly but when I do I get several failures since Rails isn't guaranteed to exist and have a `cache` object on it in the actionview library

```
Error:
CachedCollectionViewRenderTest#test_collection_caching_with_partial_that_doesn't_use_fragment_caching:
NoMethodError: undefined method `cache' for Rails:Module
```

File: activesupport/test/cache/key_test.rb [hunk 8]
```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+require "abstract_unit"
+require "active_support/cache"
+
+class KeyTest < ActiveSupport::TestCase
```
[the remark this answers is not in the document]
Comment (schneems) [reply activesupport/test/cache/key_test.rb]: Good idea. There's a lot of other tests inside of the `activesupport/cache` folder that could benefit from this pattern. I updated my PR to implement your suggestion. Resolving this comment.

File: activesupport/lib/active_support/cache/key.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (kaspth) [reply activesupport/lib/active_support/cache/key.rb]: You shouldn't call `Rails.cache` there, we've already set up a `collection_cache` to use there which fetches what's basically `Rails.cache` via the Railtie: https://github.com/rails/rails/blob/042b6fec877582f13775903f64d2c5e1da485803/actionview/lib/action_view/railtie.rb#L96

We should convert this to `mattr_accessor` ala https://github.com/rails/rails/blob/042b6fec877582f13775903f64d2c5e1da485803/actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb#L10 and remove the default_cache_object method.

File: activesupport/lib/active_support/cache/key.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+      def cache_key
+        @cache_key ||= @cache_object.send(:expanded_key, @key_parts)
+      end
+
+      def cache_version
+        @cache_version ||= @cache_object.send(:expanded_version, @key_parts)
+      end
+
+      def cache_key_with_version
+        @cache_object.send(:retrieve_cache_key, @key_parts)
```
Comment (kaspth) [inline activesupport/lib/active_support/cache/key.rb]: Not too keen on the `_object` variable suffix since everything is an object in Ruby, so it doesn't signify much 😄 How about `@cache_store`?

File: actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb [hunk 10]
```diff
@@ -59,17 +59,23 @@ def collection_by_cache_keys(view, template)
...
         digest_path = view.digest_path_from_template(template)
+        view_parts = view.combined_fragment_cache_key(nil)
+        base = view_parts.shift
 
         @collection.each_with_object([{}, []]) do |item, (hash, ordered_keys)|
-          key = expanded_cache_key(seed.call(item), view, template, digest_path)
+          key = ActiveSupport::Cache::Key.new(base)
+          view_parts.each do |part|
+            key << part
+          end
```
Comment (kaspth) [inline actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb]: Could add a `concat` method to `Key` that delegates to `concat` on `@key_parts` and do `key.concat view_parts` should be clearer and quicker still.

File: actionview/lib/action_view/helpers/cache_helper.rb [hunk 11]
```diff
@@ -227,16 +228,22 @@ def digest_path_from_template(template) # :nodoc:
...
+          if container
+            container << digest_path
+            container << name
+          else
+            [ digest_path, name ]
+          end
         else
+          container << name if container
           name
         end
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/cache_helper.rb]: Don't think this is worth saving the extra array allocation. Curious if we could replace all this with something like:

```ruby
view.cache_fragment_name(item, virtual_path: template.virtual_path, digest_path: digest_path).tap do |key_parts|
  key.concat Array(key_parts)
end
```

in the collection caching stuff. Fine if you want to replace `Array()` with more granular handling.

**Review by kaspth — COMMENTED** [review]
Getting closer!

**Comment by schneems** [comment]
Thanks for the comments I've not been able to re-visit this PR.

As a data point, I did some benchmarking with Derailed against CodeTriage and I was able to make the homepage 1.5x faster by disabling collection cache. I don't think even my patches would have been enough to break even since the partials were so small https://github.com/codetriage/codetriage/pull/1226.

While I do think we can work to make the collection cache code faster, i'm also thinking that the bigger problem people are likely having is that they're using collection caching for tiny partials when they shouldn't.
